### PR TITLE
Extract repeated logic to traits

### LINKED ISSUE #2373: PHP Refactoring to facilitate testing and 3P plugin integrations
## Feature Description

When working on #113 and #1117 it has become apparent that the PHP refactoring originally tentatively planned for post-stable  is becoming more urgent.

We currently use a colorful mix of static class methods, and classes instantiated at various stages without the possibility to easily modify their behavior.

Example:

For #113, we'd need to be able to remove all the filters added by the `Discovery` class. However, this is currently not easily possible.

This is a good opportunity to reduce technical debt on the PHP side of things.

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

* More robust but not overly complex plugin architecture  
  Ideally first drafted in a design doc or proof-of-concept PR
* 3P plugin integrations are more easily possible

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary
Move much of the repeated code to PHP trails. This will in hope make the code easier to test and refactor. 

3 of 3. 

## Relevant Technical Choices

I discussed not using trail, but just using php functions. However, trails a much clearer and keep classes single purpose. 

There are some other tweaks and tidy ups found in this PR. If this is a problem can be moved out of this PR. 
## Testing Instructions

---

Fixes #2373

### CODE REVIEW

File: includes/Dashboard.php [hunk 1]
```diff
@@ -28,12 +28,15 @@
 
 namespace Google\Web_Stories;
 
+use Google\Web_Stories\TRAITS\Assets;
```
Comment (swissspidy) [inline includes/Dashboard.php]: Can we use `Traits` instead of `TRAITS` everywhere?

File: includes/TRAITS/Assets.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * @package Google\Web_Stories\TRAITS
+ */
+trait Assets {
+	/**
+	 * Helper method to load assets based on a handle.
+	 *
+	 * @param string $handle Handle of script.
+	 * @param array  $extra_dependencies Array of extra dependencies.
+	 */
+	public function load_assert( $handle, array $extra_dependencies = [] ) {
```
Comment (swissspidy) [inline includes/TRAITS/Assets.php]: Typo: `assert` -> `asset`

File: includes/TRAITS/Assets.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * Trait Assets
+ *
+ * @package Google\Web_Stories\TRAITS
+ */
+trait Assets {
+	/**
+	 * Helper method to load assets based on a handle.
+	 *
+	 * @param string $handle Handle of script.
+	 * @param array  $extra_dependencies Array of extra dependencies.
```
Comment (swissspidy) [inline includes/TRAITS/Assets.php]:
```suggestion
	 * @param array  $extra_dependencies Array of extra dependencies.
	 *
	 * @return void
```

File: includes/TRAITS/Publisher.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+		// the editor should take care of warning about incorrect size.
+		return is_array( $fullsize_img ) ? array_shift( $fullsize_img ) : false;
+	}
+
+	/**
+	 * Helper function to get option name.
+	 *
+	 * @return string
+	 */
+	public function get_publisher_logo_option() {
```
Comment (swissspidy) [inline includes/TRAITS/Publisher.php]: Maybe `get_publisher_logo_option_name`? Otherwise a bit confusing

File: includes/TRAITS/Assets.php (hunk 2 above)
Comment (swissspidy) [reply includes/TRAITS/Assets.php]: Should we rename this to `load_script`? And create `load_style` in addition?

File: includes/Dashboard.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Dashboard.php]: fixed.

File: includes/TRAITS/Publisher.php (hunk 4 above)
Comment (spacedmonkey) [reply includes/TRAITS/Publisher.php]: Fixed.

File: includes/TRAITS/Assets.php (hunk 3 above)
Comment (spacedmonkey) [reply includes/TRAITS/Assets.php]: Fixed

File: includes/TRAITS/Assets.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/TRAITS/Assets.php]: I have changed to `load_asset`. This one method now loads css and js.

File: includes/Embed_Block.php [hunk 5]
```diff
@@ -53,27 +57,7 @@ public function init() {
...
-			false
-		);
-
-		wp_register_style(
-			self::STYLE_HANDLE,
-			WEBSTORIES_PLUGIN_DIR_URL . 'assets/css/' . self::STYLE_HANDLE . '.css',
-			[ 'amp-story-player' ],
-			$version
-		);
+		$this->load_asset( self::SCRIPT_HANDLE, [ 'amp-story-player' ], self::STYLE_HANDLE, [ 'amp-story-player' ] );
```
Comment (swissspidy) [inline includes/Embed_Block.php]: 🚨 This now calls `wp_enqueue_script`. What we want & need here is `wp_register_script`.

There needs to be a `register_asset` method for that.

File: includes/TRAITS/Assets.php (hunk 2 above)
Comment (swissspidy) [reply includes/TRAITS/Assets.php]: What if we ever want to load just CSS, no JS?

Also, as per my other comment we need to make it possible to just _register_ assets, not _enqueue_.

Wouldn't be easier if we'd have a a structure like this:

```php
private function get_asset_metadata( $handle ) {
	$asset_file   = WEBSTORIES_PLUGIN_DIR_PATH . 'assets/js/' . $script_handle . '.asset.php';
	$asset        = is_readable( $asset_file ) ? require $asset_file : [];
	if ( ! $asset['dependencies'] ){
		$asset['dependencies'] = [];
	}
	if ( ! $asset['version'] ){
		$asset['version'] = WEBSTORIES_VERSION;
	}

	return $asset;
}

public function register_script( $handle, $extra_dependencies ) {
	$metadata = $this->get_asset_metadata(...);
	wp_register_script(...);
	wp_set_script_translations( $handle, 'web-stories' );
}

public function enqueue_script( $handle, $extra_dependencies ) {
	$this->register_script( $handle, $extra_dependencies );
	wp_enqueue_script( $handle );
}

public function register_style( $handle, $extra_dependencies ) {
	$metadata = $this->get_asset_metadata(...);
	wp_register_style(...);
}

public function enqueue_style( $handle, $extra_dependencies ) {
	$this->register_style( $handle, $extra_dependencies );
	wp_enqueue_style( $handle );
}
```

File: includes/Embed_Block.php (hunk 5 above)
Comment (spacedmonkey) [reply includes/Embed_Block.php]: It was using `wp_enqueue_script` before. See https://github.com/google/web-stories-wp/blob/c15743f20af24f945391fde41d3a5aa6e33f1159/includes/Story_Post_Type.php#L285-L291

Comment (swissspidy) [reply includes/Embed_Block.php]: That's in `Story_Post_Type.php`. This here is in `Embed_Block.php`

File: includes/Traits/Url_Parser.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+use DOMXpath;
+use DOMNodeList;
+use DOMElement;
+
+/**
+ * Trait Url_Parser
+ *
+ * @package Google\Web_Stories\Traits
+ */
+trait Url_Parser {
```
Comment (swissspidy) [inline includes/Traits/Url_Parser.php]: Maybe `Document_Parser`?

File: includes/TRAITS/Assets.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/TRAITS/Assets.php]: Very valid change. I like having all scripts and styles registered.

File: includes/Embed_Block.php (hunk 5 above)
Comment (spacedmonkey) [reply includes/Embed_Block.php]: Fixed.

File: includes/Traits/Url_Parser.php (hunk 6 above)
Comment (spacedmonkey) [reply includes/Traits/Url_Parser.php]: Changed

File: includes/REST_API/Link_Controller.php [hunk 7]
```diff
@@ -105,7 +105,7 @@ public function parse_link( $request ) {
 
 		$data = get_transient( $cache_key );
 		if ( ! empty( $data ) ) {
-			return json_decode( $data, true );
+			return rest_ensure_response( json_decode( $data, true ) );
```
Comment (spacedmonkey) [inline includes/REST_API/Link_Controller.php:108]: This is a fix! @swissspidy

File: includes/Dashboard.php [hunk 8]
```diff
@@ -145,21 +148,35 @@ public function enqueue_assets( $hook_suffix ) {
...
+			[ 'forms' ]
+		);
+	}
 
+	/**
+	 * Get dashboard settings as an array.
+	 *
+	 * @return array
+	 */
+	public function get_dashboard_settings() {
```
Comment (spacedmonkey) [inline includes/Dashboard.php:176]: Move this to it's own method to make this a little cleaner.

File: includes/Traits/Assets.php [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	 */
+	public function enqueue_style( $style_handle, array $style_dependencies = [] ) {
+		$this->register_style( $style_handle, $style_dependencies );
+		wp_enqueue_style( $style_handle );
+	}
+
+	/**
+	 * Remove admin styles.
+	 *
+	 * @param array $styles Array to style to be removed.
```
Comment (swissspidy) [inline includes/Traits/Assets.php:123]:
```suggestion
	 * @param array $styles Array to style to be removed.
	 * @return void
```

File: includes/Traits/Document_Parser.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace Google\Web_Stories\Traits;
+
+use DOMDocument;
+use DOMXpath;
+use DOMNodeList;
+use DOMElement;
+
+/**
+ * Trait Url_Parser
```
Comment (swissspidy) [inline includes/Traits/Document_Parser.php]:
```suggestion
 * Trait Document_Parser
```

File: includes/Traits/Document_Parser.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Trait Url_Parser
```
Comment (swissspidy) [inline includes/Traits/Document_Parser.php]:
```suggestion
 * Trait Document_Parser
```

File: includes/Story_Renderer.php [hunk 12]
```diff
@@ -236,4 +234,28 @@ public function render() {
...
+
+	/**
+	 * Get story meta images.
+	 *
+	 * There is a fallback poster-portrait image added via a filter, in case there's no featured image.
+	 *
+	 * @param int|\WP_Post|null $post Post.
+	 * @return string[] Images.
+	 */
+	public function get_story_meta_images( $post = null ) {
```
Comment (swissspidy) [inline includes/Story_Renderer.php]: Can this be kept `protected`? `$post` can then be _required_.

Comment (spacedmonkey) [reply includes/Story_Renderer.php]: This one was a copy and paste from Media class. But I will make the changes.

Comment (spacedmonkey) [reply includes/Story_Renderer.php]: Fixed in https://github.com/google/web-stories-wp/pull/2623/commits/f09fe8405bc103857762050794b506a3657fead8

Comment (spacedmonkey) [reply includes/Story_Renderer.php]: Fixed.

**Review by swissspidy — APPROVED** [review]

**Pull request merged**